### DISCUSSION TITLE
Crap GPS-Signal detection

### DISCUSSION THREAD

**Discussion by falumas** [time4tea/gopro-dashboard-overlay#88] [Ideas]
When I had bad GPS reception during the recording the overlay goes bananas (not a fault by this software).

Would it be possible to add a min and max speed and if the speed is outside those limits skip adding the overlay for those frames. 

I am not the worst person who ever went down a slalom slope, but the 30'000kmh that I sometimes get on the overlay might be a bit exaggerated. My experience after that I have processed 40-50 videos, where maybe 10 had crap telementary, is that once the speed comes below 100kmh it fairly quickly after that starts to be good. Or it just stays at 0 if no signal was picked up.

**Comment by armanddp** [comment]
There's a precision (or rather lack of) score in the GoPro (gpmd) metadata. Which should be quite useful to filter out the bad readings. While I couldn't find a reference to it quickly scanning I was pretty sure it might already be in use on this great project. 


`GPSP | GPS Precision - Dilution of Precision (DOP x100) | 1 | n/a | Within the GPS stream, under 500 is good` - https://github.com/gopro/gpmf-parser

GPSP (or DOP) see https://en.wikipedia.org/wiki/Dilution_of_precision_(navigation)

**Reply by time4tea** [reply]
Hiya. Thanks for using the software! We do parse the DOP value and theoretically we could use it as a metric value, although to be honest I don't think I've implemented that..

In older versions of the software I did filter out values with a 'bad' DOP, but I removed this when making some changes to the parsing code.

It wasn't clear to me what the correct behaviour should be..

There are some things that we could do, one or more of

- filter out gps points with DOP > some value
- allow command line selection of max allowable DOP
- add metric / icon to show DOP
- add icon to show GPS acquisition

The software probably doesn't do sensible things right now if the GPS time or location jumps large distances. It is best to wait for signal acquire before setting off, if possible.

The GPS track can be very bad! I've seen the GPS move hundreds of metres into the sea while claiming to be locked. This is a firmware thing, but GoPro won't be making new firmware for older units. I have a Hero9, it often goes wandering around. I will be getting a Hero11 soon, they claim the GPS tracking is improved.

You can use a garmin GPS track to replace the gopro one, (if you use both gadgets at the same time), using `--gpx` or `--fit` - this is likely to be much more accurate.

**Reply by time4tea** [reply]
Sorry I re-read your comment about max speed filter - yes,  could do this too, but we'd need to be careful on how to set the defaults, as people put their gopro in planes too! 😀

All good suggestions, thank you!

**Reply by armanddp** [reply]
The Hero11 (and 10) with the newer firmware is so much better. Slightly slower 10hz GPS updates (from 18hz) but with much better accuracy. Hero9 is terrible, I've also recently been testing with one and it's all over the place, mostly unusable! 

- allow command line selection of max allowable DOP 
- add metric / icon to show DOP

My 2c would be they seem to be most flexible. Working on something similar and that's the direction we're leaning. Give the option to show that the GPS is bad, it's not you cycling into the ocean :)

**Reply by time4tea** [reply]
@falumas - one thing that is really helpful for me is if you are able to share a video where it all goes wrong. I can then see exactly what happens in the telemetry stream. If you are able to do this please use WeTransfer or similar to send to gopro-overlay [at] time4tea.net . Please don't feel obliged to though. All your data will remain private.

**Reply by falumas** [reply]
If possible the solution suggested by @armanddp looks much better and that will allow the people on ISS to still use this awesome software as well.

Parameters that allow to set limits on these two values would be much better, if the data quality of those values are high.
GPSF | GPS Fix | 1 | n/a | Within the GPS stream: 0 - no lock, 2 or 3 - 2D or 3D Lock
GPSP | GPS Precision - Dilution of Precision (DOP x100) | 1 | n/a | Within the GPS stream, under 500 is good

I will send you a file later that contains a really bad signal the first half of the movie, but gets good enough GPS signal second half to create a feasible overlay.

**Reply by KyleGW** [reply]
I like the options presented above a lot. At one point for a video I really wanted to get right, that had bad GPS data, I exported it to a GPX, did some analysis in excel to see what the right values were for "not reasonable distance to have traveled based on the points around it", and then used that to iterate through using python and throw all the unreasonable values out. Having the max DOP used would be great. Even just a constant in the code somewhere that can be changed by advanced users  would be a good start. (But then if going that far, not much more work to make it another command line option to change it)

**Reply by time4tea** [reply]
Good to know! If something is annoying, please do raise an issue to let me know... then I can do something about it.
It might not be possible to address every single thing... but at least I can think about it :-)

**Reply by falumas** [reply]
I have sent you an email to a slightly modified adress of gopro-overlay [at] time4tea.net with download Links for the file that has a mix of data quality of the GPS Signal. Please give me a notification once you have downloaded them so that I can remove them.

**Reply by time4tea** [reply]
@falumas hiya - thank you - yes got your email - downloading them... will take a while as my internet is not that fast..

**Comment by time4tea** [comment]
@falumas @armanddp @KyleGW  - you should find some changes in latest version 0.77.0 that improve the situation a bit. please let me know if they are useful for you.

**Reply by falumas** [reply]
It works perfectly for the speedometer! Many thanks.

Is it possible to also sort bad telementary data out from
<img width="164" alt="image" src="https://user-images.githubusercontent.com/92432331/212835760-33e7fdc6-273c-43d2-94f9-7fdf708610c9.png">
and
<img width="229" alt="image" src="https://user-images.githubusercontent.com/92432331/212835804-f4d93fbc-006d-4d03-a5a9-96d95c45405e.png">

**Reply by time4tea** [reply]
Hiya. Great! For the altitude, yes. For the route... well its a bit more difficult, as I'm already removing all the points that are "not locked" and also using a mechanism to try to determine which other points, that are marked as locked, are actually not. It's a bit.of a guessing game. I can see if there is something obvious about the data, so I can filter it out, but I'm not sure right now.

I'll see what I can come up with.

To see the actual data, by the way, you can use gopro-to-gpx and then load the gpx file into a gpx map on the Web... you.can see that a lot of bad data is already filtered...

**Reply by time4tea** [reply]
@falumas maybe the changes in the latest version will make a difference? 0.79.0... please let me know if the changes in that release were useful for you.

There are now a number of controls to clip gps  points that fail certain tests - e.g. too fast, high DOP, or outside bounding box. 

The chart component or a metric field can show the DOP as an overlay on the movie, or you can use gopro-to-csv to see this in a csv file.

**Reply by falumas** [reply]
I am trying it out, some of my video seems though to have so crappy data that nothing will help. A GoPro 11 is heading my way, hopefully that one will produce better GPS data. These changed did though help me for plenty of videos.
Once again a big thanks for your help.

**Reply by falumas** [reply]
I have now testet it quite a bit and it is working fine. The new parameters are very helpful
--gps-speed-max --gps-speed-max-units --gps-dop-max

The only thing I would change is the default value for max speed which seems to be 70kmh at the moment. 500kmh (which is enough for most landbased stuff) or so is probably a better default value. Both for skiing and bicycling 70kmh is too low, atleast here in Switzerland where you either go 5kmh upwards or 80kmh downward if you are using something that glides or rolls.

By crappy GPS-data I mainly get over 10000kmh.

**Reply by time4tea** [reply]
thanks for letting me know - that is really useful! - I'll raise the limit... I'm glad to know that the new options are working as they should.

If you have a frame with an overlay  that I could add to the examples that would be amazing - if you are able, please send to gopro-overlay [at] time4tea.net and I'll add it. Of course, please don't feel obliged.
